### PR TITLE
Publish C++ developer docs

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR adds the developer documentation to our doxygen. The changes to enable this are minimal: the files have been moved from cpp/docs to cpp/doxygen and added to the `INPUT` section of the Doxyfile, and a custom `DoxygenLayout.xml` has been added to include the necessary header (note that this file was autogenerated with `doxygen -l` and the only modification was adding one tab for the developer guide). I added an anchor to the main developer guide header so that it can be linked to from the header.

Our current developer guide was written to be viewed on Github. As a result, it uses some [Github Flavored Markdown](https://github.github.com/gfm/) extensions, primarily around the way that code is displayed. Since doxygen does not support those, I had to make some modifications to the contents of the docs so that they would render the same way. Some care is needed around escaping the `@` symbol in the right locations in the docs to prevent doxygen from interpreting commands in examples. Finally due to https://github.com/doxygen/doxygen/issues/6054 we cannot put certain commands inside code blocks and get them rendered correctly. However, with a few cycles of rebuilding and checking the output I was able to get everything to look correct. These changes mean that the guide will no longer look as nice as before when viewed directly on Github. However, the goal of this PR is to allow everyone to move towards viewing the built and published documentation instead, so this isn't a problem.

This is how it looks now:

image

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -181,11 +179,9 @@ detail see the [`libcudf++` presentation.](https://docs.google.com/presentation/
 libcudf functions typically take views as input (`column_view` or `table_view`)
 and produce `unique_ptr`s to owning objects as output. For example,
 
-```c++
-std::unique_ptr<table> sort(table_view const& input);
-```
```
Comment (PointKernel) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:186]: Is there a way to preserve the cpp code markdown for github view and ignore it when generating docs with Doxygen?

Not a big issue but those changes will make the doc less readable from Github

Comment (vyasr) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md]: OK I was able to find a reasonable solution here. I've configured doxygen to preprocess MD files and replace the Github-specific code fences with ones that doxygen can interpret. I can't get this to work nicely in the `DOCUMENTATION.md` file because no matter what I do doxygen tries to interpret doxygen commands inside code fences in MD files, but it works for `DEVELOPER_GUIDE.md` and `TESTING.md` which I think should address the vast majority of cases.

Comment (PointKernel) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md]: > If you have any idea if there's a way to change that behavior selectively

Not yet, but we will face the same issue in cuco. Will come back if we find a solution there.

> The other files have nice highlighting now

:fire: I wasn't even sure how to do it when asking the question. Thanks for making it work! Great job!

**Review by PointKernel — APPROVED** [review]

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# libcudf C++ Developer Guide
+# libcudf C++ Developer Guide {#DEVELOPER_GUIDE}
 
 This document serves as a guide for contributors to libcudf C++ code. Developers should also refer
 to these additional files for further documentation of libcudf best practices.
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:4]: Do the links to "these additional files" work? Should this top section be rewritten for a web-based audience?

File: cpp/doxygen/developer_guide/DOCUMENTATION.md [hunk 3]
```diff
@@ -246,156 +234,136 @@ The following tags should appear near the end of function comment block in the o
...
+| [\@throw](#throw) | Specify the conditions in which the function may throw an exception |
+| [\@tparam](#tparam) | Description for each template parameter |
+| [\@param](#param) | Description for each function parameter |
+| [\@return](#return) | Short description of object or value returned |
 
-#### @throw
+#### \@throw
 
-Add an [@throw](http://www.doxygen.nl/manual/commands.html#cmdthrow) comment line in the doxygen block for each exception that the function may throw.
+Add an [\@throw](http://www.doxygen.nl/manual/commands.html#cmdthrow) comment line in the doxygen block for each exception that the function may throw.
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DOCUMENTATION.md:244]:
```suggestion
Add an [\@throw](https://www.doxygen.nl/manual/commands.html#cmdthrow) comment line in the doxygen block for each exception that the function may throw.
```

File: cpp/doxygen/developer_guide/DOCUMENTATION.md [hunk 4]
```diff
@@ -198,45 +192,39 @@ Also, try to include a short [example](#inline-examples) if possible.
...
-/**
- * @copydoc cudf::segmented_count_set_bits(bitmask_type const*,std::vector<size_type> const&)
- *
- * @param[in] stream Optional CUDA stream on which to execute kernels
- */
-std::vector<size_type> segmented_count_set_bits(bitmask_type const* bitmask,
-                                                std::vector<size_type> const& indices,
-                                                rmm::cuda_stream_view stream = cudf::default_stream_value);
-```
+You can use the [\@copydoc](http://www.doxygen.nl/manual/commands.html#cmdcopydoc) tag to avoid duplicating the comment block for a function definition.
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DOCUMENTATION.md:210]:
```suggestion
You can use the [\@copydoc](https://www.doxygen.nl/manual/commands.html#cmdcopydoc) tag to avoid duplicating the comment block for a function definition.
```

File: cpp/doxygen/developer_guide/DOCUMENTATION.md [hunk 5]
```diff
@@ -198,45 +192,39 @@ Also, try to include a short [example](#inline-examples) if possible.
 
 ### @brief
 
-The `@brief` text should be a short, one sentence description.
+The [\@brief](http://www.doxygen.nl/manual/commands.html#cmdbrief) text should be a short, one sentence description.
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DOCUMENTATION.md:195]:
```suggestion
The [\@brief](https://www.doxygen.nl/manual/commands.html#cmdbrief) text should be a short, one sentence description.
```

File: cpp/doxygen/developer_guide/DOCUMENTATION.md [hunk 6]
```diff
@@ -246,156 +234,136 @@ The following tags should appear near the end of function comment block in the o
...
+     * @throw cudf::logic_error if `input_argument` is negative or zero
+     *
 
-Using `@throws` is also acceptable but vs-code and other tools only do syntax highlighting on `@throw`.
+Using \@throws is also acceptable but VS Code and other tools only do syntax highlighting on \@throw.
 
 #### @tparam
 
-Add a [@tparam](http://www.doxygen.nl/manual/commands.html#cmdtparam) comment line for each template parameter declared by this function.
+Add a [\@tparam](http://www.doxygen.nl/manual/commands.html#cmdtparam) comment line for each template parameter declared by this function.
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DOCUMENTATION.md:258]: Use `https:` everywhere.

File: cpp/doxygen/developer_guide/DOCUMENTATION.md [hunk 7]
```diff
@@ -246,156 +234,136 @@ The following tags should appear near the end of function comment block in the o
...
-```
+     *
+     * @code
+     * auto result1 = make_column( ); // reference link will not be created
+     * auto result2 = cudf::make_column( ); // reference link will be created
+     * @endcode
+     *
 
-Although using 3 backtick marks `` ``` `` for example blocks will work too, they do not standout as well in vs-code and other source editors.
+Although using 3 backtick marks \`\`\` for example blocks will work too, they do not standout as well in VS Code and other source editors.
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DOCUMENTATION.md:339]:
```suggestion
Although using 3 backtick marks \`\`\` for example blocks will work too, they do not stand out as well in VS Code and other source editors.
```

File: cpp/doxygen/developer_guide/DOCUMENTATION.md [hunk 8]
```diff
@@ -415,48 +383,44 @@ The [doxygen_groups.h](../include/doxygen_groups.h) file does not need to be inc
...
+     * @ingroup transformation_fill
+     *
+     * @param ...
+     * @return ...
+     */
+    std::unique_ptr<column> fill(table_view const& input,...);
+
+    }  // namespace cudf
+
+You can also use the \@addtogroup with a `@{ ... @}` pair to automatically include  doxygen comment blocks as part of a group.
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DOCUMENTATION.md:402]:
```suggestion
You can also use the \@addtogroup with a `@{ ... @}` pair to automatically include doxygen comment blocks as part of a group.
```

File: cpp/doxygen/developer_guide/TESTING.md [hunk 9]
```diff
@@ -79,6 +78,7 @@ the same tests across multiple types, we use GTest's
 Typed tests allow you to write a test once and run it across a list of types.
 
 For example:
+
 ```c++
```
Comment (bdice) [inline cpp/doxygen/developer_guide/TESTING.md:82]: I'm confused about what cases are supposed to use a raw indent to indicate "code block" and what cases are supposed to use triple-backticks. The triple backticks are my preference. Does Doxygen assume the language to use for syntax highlighting?

File: cpp/doxygen/developer_guide/TESTING.md [hunk 10]
```diff
@@ -91,7 +91,7 @@ TYPED_TEST(TypedTestFixture, FirstTest){
 }
 ```
 
-To specify the list of types to use, instead of GTest's `::testing::Types<...>`, libcudf provides `cudf::test::Types<...>` which is a custom, drop-in replacement for `::testing::Types`.
+To specify the list of types to use, instead of GTest's `testing::Types<...>`, libcudf provides `cudf::test::Types<...>` which is a custom, drop-in replacement for `testing::Types`.
```
Comment (bdice) [inline cpp/doxygen/developer_guide/TESTING.md:94]: I'm a little confused about why the changes to the `::testing` namespace were needed. The code appears to mix styles between `::testing` and `testing`.

File: cpp/doxygen/developer_guide/TESTING.md [hunk 11]
```diff
@@ -213,18 +213,19 @@ columns, an additional iterator can be provided to indicate the validity of each
...
-auto elements = make_counting_transform_iterator(0, [](auto i){ return i; });
-auto validity = make_counting_transform_iterator(0, [](auto i){ return i%2; });
-fixed_point_column_wrapper<int32_t> w(elements, elements + 5, validity, 2);
+    // Creates a non-nullable column of 4 DECIMAL32 elements of scale 3: {1000, 2000, 3000, 4000}
+    auto elements = make_counting_transform_iterator(0, [](auto i){ return i; });
+    fixed_point_column_wrapper<int32_t> w(elements, elements + 4, 3);
+
+    // Creates a nullable column of 5 DECIMAL32 elements of scale 2: {null, 100, null, 300, null}
+    auto elements = make_counting_transform_iterator(0, [](auto i){ return i; });
+    auto validity = make_counting_transform_iterator(0, [](auto i){ return i%2; });
```
Comment (bdice) [inline cpp/doxygen/developer_guide/TESTING.md:224]:
```suggestion
    auto validity = make_counting_transform_iterator(0, [](auto i){ return i % 2; });
```

File: cpp/doxygen/modify_fences.sh [hunk 12]
```diff
@@ -0,0 +1,9 @@
+#~/bin/bash
```
Comment (bdice) [inline cpp/doxygen/modify_fences.sh:1]: Typo?
```suggestion
#!/bin/bash
```

File: cpp/doxygen/modify_fences.sh [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+# This script modifies the Github Markdown style code fences in our MD files
```
Comment (bdice) [inline cpp/doxygen/modify_fences.sh:5]:
```suggestion
# This script modifies the GitHub Markdown style code fences in our MD files
```

File: cpp/doxygen/modify_fences.sh [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+# into the PHP style that Doxygen supports, allowing us to display code
+# properly both on the Github GUI and in published Doxygen documentation.
```
Comment (bdice) [inline cpp/doxygen/modify_fences.sh:7]:
```suggestion
# properly both on the GitHub GUI and in published Doxygen documentation.
```

**Review by bdice — CHANGES_REQUESTED** [review]

File: cpp/doxygen/developer_guide/TESTING.md (hunk 9 above)
Comment (PointKernel) [reply cpp/doxygen/developer_guide/TESTING.md:82]: > Does Doxygen assume the language to use for syntax highlighting?

I don't think so.

The above change of replacing backticks with indent might be a leftover.

File: cpp/doxygen/developer_guide/TESTING.md (hunk 10 above)
Comment (harrism) [reply cpp/doxygen/developer_guide/TESTING.md:94]: It's normal in gtest code to use `::testing::...`  This is the style used in the gtest documentation. I'm not sure why, though...

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md (hunk 2 above)
Comment (vyasr) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:4]: Yes, these links work.

File: cpp/doxygen/developer_guide/TESTING.md (hunk 9 above)
Comment (vyasr) [reply cpp/doxygen/developer_guide/TESTING.md:82]: https://github.com/rapidsai/cudf/pull/11475#discussion_r957902484

File: cpp/doxygen/developer_guide/TESTING.md (hunk 10 above)
Comment (vyasr) [reply cpp/doxygen/developer_guide/TESTING.md:94]: My guess is it's a matter of being explicit and avoiding namespace clashes since it's probably quite common for other projects making use of gtest to have their own `testing` namespaces. The leading `::` guarantees global scope resolution.

As for why I had to change it, I am not entirely sure what doxygen's scope resolution rules are but in some cases (like this one) the leading `::` leads to doxygen failing to find the right link. I get errors like this from the pre-commit doxygen check:
```
cudf/cpp/doxygen/developer_guide/TESTING.md:67: warning: explicit link request to 'testing::Test' could not be resolved
```
I chose to remove the leading colons for expediency to get past this issue since I really have no idea how to resolve it more generally. My best guess is that the global namespace in the context of this document is some doxygen namespace that doesn't contain anything, so the name lookup fails. I don't know how else to resolve this, and I think it's a minor enough issue that it's OK to move forward anyway. If you can find a better solution I'm all for it, though.

File: cpp/doxygen/modify_fences.sh (hunk 12 above)
Comment (vyasr) [reply cpp/doxygen/modify_fences.sh:1]: Yup thanks.

File: cpp/doxygen/developer_guide/TESTING.md (hunk 9 above)
Comment (vyasr) [reply cpp/doxygen/developer_guide/TESTING.md:82]: In summary: I also prefer the backticks, but I used the indented code blocks in DOCUMENTATION.md so that doxygen wouldn't interpret doxygen commands inside the Markdown file.

Comment (bdice) [reply cpp/doxygen/developer_guide/TESTING.md:82]: Thanks. That's fine.

File: cpp/doxygen/developer_guide/TESTING.md (hunk 10 above)
Comment (bdice) [reply cpp/doxygen/developer_guide/TESTING.md:94]: I don't want to hold up the PR on this but let's leave this conversation unresolved.

File: cpp/doxygen/Doxyfile [hunk 15]
```diff
@@ -962,7 +966,7 @@ IMAGE_PATH             =
 # need to set EXTENSION_MAPPING for the extension otherwise the files are not
 # properly processed by doxygen.
 
-INPUT_FILTER           =
+INPUT_FILTER           =
```
Comment (bdice) [inline cpp/doxygen/Doxyfile:969]: Is this trailing whitespace required? If not, remove.
```suggestion
INPUT_FILTER           =
```

**Review by bdice — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge

**Comment by vyasr** [comment]
rerun tests
